### DISCUSSION TITLE
Dynamic This PC drive list creation

### DISCUSSION THREAD

**Discussion by Assimilator-X** [moudey/Shell#582] [General]
How do i go about to create a dynamic drive list menu of This PC? I am on Windows 11 Enterprise LTSC,  26100.2605. 
Trying to create menu that lists my drives. So far i have tried,

	menu( mode='multiple' title='This PC' image=\uE1D9)
{
		Item (title = 'A: WD120EDBZ' where=!path.exists('A:') image)
		Item (title = 'B: WD120EDAZ' where=!path.exists('B:') image)
}
 Nothing comes up in the menu also need help to apply the vis= parameter so its only listing the item if its exist.
 Thanks.
<img width="1920" alt="2024-12-19_225441" src="https://github.com/user-attachments/assets/ebe85929-47b0-4a0d-82de-d5fbd48ea61c" />
Would it be even possible or i just poking around in the dark? LOL

**Comment by RubicBG** [comment]
does this help ypu?
```js
menu( mode='multiple' title='This PC' image=\uE1D9)
{
	item (title = 'A:' where=path.exists('A:') image=icon.res('imageres.dll', if('A:'== sys.root, 31, 27)) vis=if('A:'== sys.root, 'disable') )
	item (title = 'B:' where=path.exists('B:') image=icon.res('imageres.dll', if('B:'== sys.root, 31, 27)) vis=if('B:'== sys.root, 'disable') )
	item (title = 'C:' where=path.exists('C:') image=icon.res('imageres.dll', if('C:'== sys.root, 31, 27)) vis=if('C:'== sys.root, 'disable') )
	item (title = 'D:' where=path.exists('D:') image=icon.res('imageres.dll', if('D:'== sys.root, 31, 27)) vis=if('D:'== sys.root, 'disable') )
	// ...
	item (title = 'X:' where=path.exists('X:') image=icon.res('imageres.dll', if('X:'== sys.root, 31, 27)) vis=if('X:'== sys.root, 'disable') )
	item (title = 'Y:' where=path.exists('Y:') image=icon.res('imageres.dll', if('Y:'== sys.root, 31, 27)) vis=if('Y:'== sys.root, 'disable') )
	item (title = 'Z:' where=path.exists('Z:') image=icon.res('imageres.dll', if('Z:'== sys.root, 31, 27)) vis=if('Z:'== sys.root, 'disable') )
}
```

**Comment by Assimilator-X** [comment]
It brings up a list yes, however when i click on a listed drive it does not open it. I was more likely looking for a similar solution that we had back in windows 7 when you put a link on the tray area of My Computer and it could dynamically list all the presend drives. Good old days when windows was customizable. LOL I dont want to overbloat the menu by adding independent sections of all my network drives and all the shared folder on all my laptops. The shell is fast as it is very fast without animations and fading effects to slow it down. 
<img width="1920" alt="2024-12-20_220837" src="https://github.com/user-attachments/assets/5e1b5a3e-ebf5-43ec-8ac3-e852614fca08" />
It might be a not existing feature for now. At least shell is much better than windows 11's menu by default. MS already blocking some customization tools by the app compatibility framework.
Thank you.

**Comment by RubicBG** [comment]
you didn't explain what you wanted to do with the menu and I didn't add any commands. now 
- if there is an open explorer it will navigate to the corresponding destination
- if there is no open window it will open one with the desired destination
- if there is an open window but you still want to open the destination in a new explorer hold shift

```js
menu(mode='multiple' title='This PC' image=\uE1D9)
{
	item (title = 'A:' where=path.exists('A:') image=icon.res('imageres.dll', if('A:'== sys.root, 31, 27)) vis=if('A:'== sys.root, 'disable') cmd=if(window.name=='CabinetWClass' and !keys.shift(), command.navigate('A:'), 'A:'))
	item (title = 'B:' where=path.exists('B:') image=icon.res('imageres.dll', if('B:'== sys.root, 31, 27)) vis=if('B:'== sys.root, 'disable') cmd=if(window.name=='CabinetWClass' and !keys.shift(), command.navigate('B:'), 'B:'))
	item (title = 'C:' where=path.exists('C:') image=icon.res('imageres.dll', if('C:'== sys.root, 31, 27)) vis=if('C:'== sys.root, 'disable') cmd=if(window.name=='CabinetWClass' and !keys.shift(), command.navigate('C:'), 'C:'))
	item (title = 'D:' where=path.exists('D:') image=icon.res('imageres.dll', if('D:'== sys.root, 31, 27)) vis=if('D:'== sys.root, 'disable') cmd=if(window.name=='CabinetWClass' and !keys.shift(), command.navigate('D:'), 'D:'))
	// ...
	item (title = 'X:' where=path.exists('X:') image=icon.res('imageres.dll', if('X:'== sys.root, 31, 27)) vis=if('X:'== sys.root, 'disable') cmd=if(window.name=='CabinetWClass' and !keys.shift(), command.navigate('X:'), 'X:'))
	item (title = 'Y:' where=path.exists('Y:') image=icon.res('imageres.dll', if('Y:'== sys.root, 31, 27)) vis=if('Y:'== sys.root, 'disable') cmd=if(window.name=='CabinetWClass' and !keys.shift(), command.navigate('Y:'), 'Y:'))
	item (title = 'Z:' where=path.exists('Z:') image=icon.res('imageres.dll', if('Z:'== sys.root, 31, 27)) vis=if('Z:'== sys.root, 'disable') cmd=if(window.name=='CabinetWClass' and !keys.shift(), command.navigate('Z:'), 'Z:'))
}
```
- `vis=if('?:'== sys.root, 'disable')` checks which hard drive is with the operating system and makes it inactive - remove it everywhere if you want to be able to open drive c: (in your case)

**Comment by Assimilator-X** [comment]
Ideally, i were trying to create a dynamic item, that cauld display the contents of <This PC> as a menu. Apologize for the lack of explanation RubicBG LOL. What i was looking for is a line and turn it into a menu without declaring all the drives that my system might or might not have and just hide them if they are not present. I was digging into the documentation dont seem to have a way to turn this line into a main menu and display its contents as a sub-menu.

menu(mode='multiple' title='This PC' image=\uE1D9)
{
item(title='This PC' image=\uE0F3 cmd='shell:::{20D04FE0-3AEA-1069-A2D8-08002B30309D}')
 }

If i do this, the whole menu crashes and dont launch.
	menu( mode='multiple' title='This PC' image=\uE1D9 cmd='shell:::{20D04FE0-3AEA-1069-A2D8-08002B30309D}')
	
{

}

Thank you.

**Comment by RubicBG** [comment]
there is no way to create items/menus recursively - it has to be done one by one
you cannot assign a command to the menu (you cannot put `cmd=` in `menu( )`)
check `shell.log` (in a text editor) to see in which file/line there is a syntax error (to avoid crashing)

**Comment by Assimilator-X** [comment]
I got to the same conclusion :( . It would have been cool.
